Platform/RaspberryPi/Library/PlatformPcdLib/PlatformPcdLib.c

```diff
@@ -0,0 +1,59 @@
+/** @file
+ *
+ *  Copyright (c) 2020, Pete Batard <pete@akeo.ie>
+ * 
+ *  SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ **/
+
+#include <Library/DebugLib.h>
+#include <Library/PcdLib.h>
+#include <Library/PrintLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Protocol/RpiFirmware.h>
+
+EFI_STATUS
+EFIAPI
+PlatformPcdLibConstructor (
+  IN EFI_HANDLE ImageHandle,
+  IN EFI_SYSTEM_TABLE *SystemTable
+  )
+{
+
+#if (FixedPcdGet64 (PcdBcmGenetRegistersAddress) != 0)
+  //
+  // If Genet is in use and a MAC address PCD has not been set, do it here.
+  //
+  EFI_STATUS                       Status;
+  UINT64                           MacAddr;
+  RASPBERRY_PI_FIRMWARE_PROTOCOL   *mFwProtocol;
+
+  Status = gBS->LocateProtocol (&gRaspberryPiFirmwareProtocolGuid, NULL,
+                  (VOID**)&mFwProtocol);
+  ASSERT_EFI_ERROR(Status);
+
+  //
+  // Get the MAC address from the firmware
+  //
+  Status = mFwProtocol->GetMacAddress ((UINT8*) &MacAddr);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_WARN, "%a: failed to retrieve MAC address\n", __FUNCTION__));

```

Comment:
Seems like this should be doing something with the value of PcdBcmGenetMacAddress in the face of failure.

Reply:
Hmm, you're right that there's an issue here.

There's an `if (PcdGet64 (PcdBcmGenetMacAddress) == 0)` that's hinted at in the comment, but that I forgot to add in this version of the code, and that should have guarded the whole MAC address retrieval and PCD updating so that we don't overwrite a PCD that's already been set by the user. I'll fix that in the code I upstream (probably tomorrow).

However, I don't see why or how we could do something with `PcdBcmGenetMacAddress` if `GetMacAddress ()` fails once we have the missing `if (PcdGet64 (PcdBcmGenetMacAddress) == 0)` guard, because then the PCD can only be zero, and we won't have a MAC to work with, so the best we could do here would be to set a random value. However, considering that the Linux drivers will detect if the UMAC has not been set and assign a random MAC value then, and I would suggest we do the same in our UEFI Genet driver (since I don't think we want to assume that UMAC has always been properly set), I'd assert that the best course of action is to just do nothing.

Or do you actually rely on the stub UMAC initialization to be performed always, even if the MAC is zero? Coz the way I saw it is that this part of the Genet code is only meant to be used for MAC setup if we have one, and that the rest of the driver would know how to deal with UMAC regardless of whether `SetMacAddress ()` has been called or not.

I guess this is something we can sort out once we start adding the actual network driver code, but it may help to clarify what you had in mind with UMAC init.

Reply:
The current code, (i'm about to post) does set a random mac if the one read from the adapter is invalid (all 0, 0xff or a multicast). So I guess its ok for the time being to have a 0.

But as you mention, that won't fly in for a future fully featured driver.